src/commands/edge_app.rs

```diff
@@ -0,0 +1,250 @@
+use crate::authentication::Authentication;
+use crate::commands::CommandError;
+use std::collections::HashMap;
+
+use log::debug;
+use reqwest::header::HeaderMap;
+use reqwest::StatusCode;
+use serde::{Deserialize, Serialize};
+use serde_yaml;
+use std::fs;
+use std::fs::File;
+use std::io::Write;
+use std::path::Path;
+
+#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct EdgeAppManifest {
+    #[serde(default)]
+    pub id: String,
+    #[serde(default)]
+    pub created_at: String,
+    #[serde(default)]
+    pub created_by: String,
+    #[serde(default)]
+    pub permissions: Vec<String>,
+    #[serde(default)]
+    pub updated_at: String,
+    pub name: String,
+    pub version: String,
+    pub description: String,
+    pub icon: String,
+    pub author: String,
+    pub homepage_url: String,
+}
+
+pub struct EdgeAppCommand {
+    authentication: Authentication,
+}
+
+impl EdgeAppCommand {
+    pub fn new(authentication: Authentication) -> Self {
+        Self { authentication }
+    }
+
+    pub fn init(self, path: &Path) -> Result<(), CommandError> {
+        let mut object = serde_yaml::to_value(EdgeAppManifest::default())?;
+        let map = object.as_mapping_mut().ok_or(CommandError::MissingField)?;
+
+        // following fields will be generated server side when publishing.
+        map.remove("id");
+        map.remove("created_at");
+        map.remove("created_by");
+        map.remove("updated_at");
+        map.remove("permissions");
+
+        let yaml = serde_yaml::to_string(map)?;
+        let input = File::create(path)?;
+        write!(&input, "{yaml}")?;
+
+        Ok(())
+    }
+
+    pub fn publish(self, path: &Path) -> Result<EdgeAppManifest, CommandError> {
+        let url = format!("{}/v4/edge_apps", &self.authentication.config.url);
+
+        let data = fs::read_to_string(path)?;
+        // by converting to struct we make sure there no extra fields and all required fields are present
+        let edge_app: EdgeAppManifest = serde_yaml::from_str(&data)?;
+        let mut binding = serde_json::to_value(edge_app)?;
```

Comment:
Not sure about conversion to json - you already have a struct edge_app - do you use it to be able to delete fields easily?

Reply:
Hmm, yeah that was the idea. But I ended up not using that code as payload is a separate struct. I will update it.

Reply:
But parsing to struct first is still valuable as it validates that all fields are present.